File: tests/wasm/test_mom.js
/**
 * WASM binding tests for MOM indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        
        const { createRequire } = await import('module');
        const require = createRequire(import.meta.url);
        wasm = require('../../pkg/my_project.js');
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        console.error('Error:', error);
        throw error;
    }
    
    testData = loadTestData();
});

test('MOM accuracy', () => {
    const period = EXPECTED_OUTPUTS.mom.defaultParams.period;
    const result = wasm.mom_js(testData.close, period);
    
    
    assert(result && (Array.isArray(result) || result instanceof Float64Array), 'MOM should return an array or typed array');
    assert.strictEqual(result.length, testData.close.length, 'Result length should match input length');
    
    
    for (let i = 0; i < period; i++) {
        assert(isNaN(result[i]), `Value at index ${i} should be NaN (warmup period)`);
    }
    
    
    const expectedLast5 = EXPECTED_OUTPUTS.mom.last5Values;
    const resultLast5 = result.slice(-5);
    
    for (let i = 0; i < 5; i++) {
        assertClose(resultLast5[i], expectedLast5[i], 0.1, `MOM last-5[${i}]`);
    }
});

test('MOM error handling', () => {
    
    assert.throws(() => {
        wasm.mom_js([], 10);
    }, /Input data slice is empty|Invalid period/);
    
    
    const nanData = new Array(10).fill(NaN);
    assert.throws(() => {
        wasm.mom_js(nanData, 5);
    }, /All values are NaN/);
    
    
    const shortData = [10.0, 20.0, 30.0];
    assert.throws(() => {
        wasm.mom_js(shortData, 10);
    }, /Invalid period/);
    
    
    assert.throws(() => {
        wasm.mom_js([10.0, 20.0, 30.0], 0);
    }, /Invalid period/);
});

test('MOM fast API (in-place)', async () => {
    const period = 10;
    
    
    const ptr = wasm.mom_alloc(testData.close.length);
    
    try {
        
        const memory = wasm.memory || wasm.__wasm.memory || wasm.wasm?.memory;
        if (!memory) {
            console.log('Skipping fast API test - memory not exposed');
            return;
        }
        const wasmMemory = new Float64Array(memory.buffer, ptr, testData.close.length);
        wasmMemory.set(testData.close);
        
        
        wasm.mom_into(ptr, ptr, testData.close.length, period);
        
        
        const result = Array.from(wasmMemory);
        
        
        const expected = wasm.mom_js(testData.close, period);
        assertArrayClose(result, Array.from(expected), 1e-10);
        
    } finally {
        
        wasm.mom_free(ptr, testData.close.length);
    }
});

test('MOM batch API', () => {
    const config = {
        period_range: [5, 15, 5]  
    };
    
    const result = wasm.mom_batch(testData.close, config);
    
    assert(result.values, 'Batch result should have values');
    assert(result.combos, 'Batch result should have combos');
    assert.strictEqual(result.rows, 3, 'Should have 3 rows for periods 5, 10, 15');
    assert.strictEqual(result.cols, testData.close.length, 'Columns should match data length');
    assert.strictEqual(result.values.length, result.rows * result.cols, 'Values array size should be rows * cols');
    
    
    const period10Result = wasm.mom_js(testData.close, 10);
    const row1 = result.values.slice(result.cols, 2 * result.cols); 
    assertArrayClose(row1, period10Result, 1e-10);
});

test('MOM partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.mom_js(close, 10); 
    assert.strictEqual(result.length, close.length, 'Result length should match input');
});

test('MOM very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.mom_js(singlePoint, 9);
    }, /Invalid period|Not enough valid data/);
});

test('MOM empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.mom_js(empty, 10);
    }, /Input data slice is empty/);
});


test('MOM NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.mom_js(close, 10);
    assert.strictEqual(result.length, close.length);
    
    
    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
    
    
    assertAllNaN(result.slice(0, 10), "Expected NaN in warmup period");
});

test('MOM all NaN input', () => {
    
    const allNaN = new Float64Array(100).fill(NaN);
    
    assert.throws(() => {
        wasm.mom_js(allNaN, 10);
    }, /All values are NaN/);
});

test('MOM warmup period', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    const period = 10;
    
    const result = wasm.mom_js(close, period);
    
    
    for (let i = 0; i < period; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} in warmup period`);
    }
    
    
    for (let i = period; i < result.length; i++) {
        assert(!isNaN(result[i]), `Unexpected NaN at index ${i} after warmup`);
    }
    
    
    
    for (let i = period; i < Math.min(period + 5, close.length); i++) {
        const expected = close[i] - close[i - period];
        assertClose(result[i], expected, 1e-10, `MOM calculation mismatch at index ${i}`);
    }
});

test('MOM batch into API', () => {
    
    const periods = [5, 10, 15];
    const rows = periods.length;
    const cols = testData.close.length;
    
    
    const inPtr = wasm.mom_alloc(cols);
    const outPtr = wasm.mom_alloc(rows * cols);
    
    try {
        
        const memory = wasm.memory || wasm.__wasm.memory || wasm.wasm?.memory;
        if (!memory) {
            console.log('Skipping batch into API test - memory not exposed');
            wasm.mom_free(inPtr, cols);
            wasm.mom_free(outPtr, rows * cols);
            return;
        }
        const wasmMemory = new Float64Array(memory.buffer, inPtr, cols);
        wasmMemory.set(testData.close);
        
        
        const resultRows = wasm.mom_batch_into(
            inPtr, outPtr, cols,
            5, 15, 5  
        );
        
        assert.strictEqual(resultRows, rows, 'Should return correct number of rows');
        
        
        const outMemory = new Float64Array(memory.buffer, outPtr, rows * cols);
        const result = Array.from(outMemory);
        
        
        const period10Result = wasm.mom_js(testData.close, 10);
        const row1 = result.slice(cols, 2 * cols);
        assertArrayClose(row1, Array.from(period10Result), 1e-10, 'Batch into row should match single computation');
        
    } finally {
        
        wasm.mom_free(inPtr, cols);
        wasm.mom_free(outPtr, rows * cols);
    }
});

test.after(() => {
    console.log('MOM WASM tests completed');
});

File: tests/python/test_mom.py
"""
Python binding tests for MOM indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path


sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'target/wheels'))

try:
    import my_project as ta_indicators
except ImportError:
    pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestMom:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_mom_partial_params(self, test_data):
        """Test MOM with default parameters - mirrors check_mom_partial_params"""
        close = test_data['close']


        result = ta_indicators.mom(close, 10)
        assert len(result) == len(close)

    def test_mom_accuracy(self, test_data):
        """Test MOM matches expected values from Rust tests - mirrors check_mom_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['mom']

        result = ta_indicators.mom(
            close,
            period=expected['default_params']['period']
        )

        assert len(result) == len(close)



        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=0,
            atol=1e-1,
            msg="MOM last 5 values mismatch"
        )

    def test_mom_default_params(self, test_data):
        """Test MOM with default parameters - mirrors check_mom_default_candles"""
        close = test_data['close']


        result = ta_indicators.mom(close, 10)
        assert len(result) == len(close)

    def test_mom_zero_period(self):
        """Test MOM fails with zero period - mirrors check_mom_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.mom(input_data, period=0)

    def test_mom_period_exceeds_length(self):
        """Test MOM fails when period exceeds data length - mirrors check_mom_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.mom(data_small, period=10)

    def test_mom_very_small_dataset(self):
        """Test MOM fails with insufficient data - mirrors check_mom_very_small_dataset"""
        single_point = np.array([42.0])

        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.mom(single_point, period=9)

    def test_mom_empty_input(self):
        """Test MOM fails with empty input - mirrors check_mom_empty_input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.mom(empty, period=10)

    def test_mom_nan_handling(self, test_data):
        """Test MOM handles NaN values correctly - mirrors check_mom_nan_handling"""
        close = test_data['close']

        result = ta_indicators.mom(close, period=10)
        assert len(result) == len(close)


        if len(result) > 240:
            assert not np.any(np.isnan(result[240:])), "Found unexpected NaN after warmup period"


        assert np.all(np.isnan(result[:10])), "Expected NaN in warmup period"

    def test_mom_streaming(self, test_data):
        """Test MOM streaming matches batch calculation - mirrors check_mom_streaming"""
        close = test_data['close']
        period = 10


        batch_result = ta_indicators.mom(close, period=period)


        stream = ta_indicators.MomStream(period=period)
        stream_values = []

        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)

        stream_values = np.array(stream_values)


        assert len(batch_result) == len(stream_values)


        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9,
                        msg=f"MOM streaming mismatch at index {i}")

    def test_mom_batch(self, test_data):
        """Test MOM batch processing with multiple periods"""
        close = test_data['close']


        result = ta_indicators.mom_batch(
            close,
            period_range=(5, 15, 5)
        )

        assert 'values' in result
        assert 'periods' in result


        assert result['values'].shape[0] == 3
        assert result['values'].shape[1] == len(close)
        assert len(result['periods']) == 3
        assert list(result['periods']) == [5, 10, 15]


        single_result = ta_indicators.mom(close, period=10)
        assert_close(
            result['values'][1],
            single_result,
            rtol=1e-10,
            msg="MOM batch period=10 mismatch with single calculation"
        )

    def test_mom_batch_single_period(self, test_data):
        """Test MOM batch with single period matches single calculation"""
        close = test_data['close']


        batch_result = ta_indicators.mom_batch(
            close,
            period_range=(10, 10, 0)
        )


        single_result = ta_indicators.mom(close, period=10)

        assert batch_result['values'].shape[0] == 1
        assert batch_result['values'].shape[1] == len(close)


        assert_close(
            batch_result['values'][0],
            single_result,
            rtol=1e-10,
            msg="MOM batch single period mismatch"
        )

    def test_mom_all_nan_input(self):
        """Test MOM with all NaN values"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.mom(all_nan, period=10)

    def test_mom_batch_edge_cases(self, test_data):
        """Test edge cases for batch processing"""
        close = test_data['close'][:50]


        result = ta_indicators.mom_batch(
            close,
            period_range=(10, 12, 10)
        )


        assert result['values'].shape[0] == 1
        assert list(result['periods']) == [10]


        with pytest.raises(ValueError):
            ta_indicators.mom_batch(
                np.array([]),
                period_range=(10, 10, 0)
            )

    def test_mom_warmup_period(self, test_data):
        """Test that MOM correctly handles warmup period"""
        close = test_data['close'][:100]
        period = 10

        result = ta_indicators.mom(close, period=period)


        assert np.all(np.isnan(result[:period])), f"Expected NaN in first {period} values"


        assert not np.any(np.isnan(result[period:])), f"Unexpected NaN after index {period}"



        for i in range(period, min(period + 5, len(close))):
            expected = close[i] - close[i - period]
            assert_close(
                result[i],
                expected,
                rtol=1e-10,
                msg=f"MOM calculation mismatch at index {i}"
            )


if __name__ == '__main__':
    pytest.main([__file__, '-v'])
